backend/src/main/java/ai/verta/modeldb/lineage/LineageServiceImpl.java

```diff
@@ -144,15 +164,92 @@ public void findAllInputsOutputs(
     }
   }
 
-  private boolean isResourceExists(Session session, String id, LineageEntryType type)
-      throws ModelDBException {
-    switch (type) {
-      case EXPERIMENT_RUN:
-        return experimentDAO.isExperimentRunExists(session, id);
-      case DATASET_VERSION:
-        return datasetVersionDAO.isDatasetVersionExists(session, id);
-      default:
-        throw new ModelDBException("Unexpected type", Code.INTERNAL);
+  private InternalFolderElementEntity getBlobFromCommit(
+      Session session, VersioningLineageEntry versioningLineageEntry) throws ModelDBException {
+    Entry<Response, InternalFolderElementEntity> result =
+        blobDAO.getCommitComponentWithHash(
+            session1 -> getRepositoryById(session, versioningLineageEntry.getRepositoryId()),
+            versioningLineageEntry.getCommitSha(),
+            versioningLineageEntry.getLocationList());
+    return result.getValue();
+  }
+
+  private RepositoryEntity getRepositoryById(Session session, long repositoryId) {
+    try {
+      return repositoryDAO.getRepositoryById(
+          session, RepositoryIdentification.newBuilder().setRepoId(repositoryId).build());
+    } catch (ModelDBException e) {
+      LOGGER.warn("repository not found {}", e.getMessage());
+      return null;
+    }
+  }
+
+  private VersioningLineageEntry getCommitFromBlob(
+      Session session, VersioningBlobElement versioningBlobElement) throws ModelDBException {
+    return blobDAO.getVersioningEntryByBlob(
+        session,
+        session1 -> getRepositoryById(session, versioningBlobElement.getRepositoryId()),
+        versioningBlobElement);
+  }
+
+  private void checkResourcesExists(Session session, List<LineageEntry> lineageEntries)
+      throws ModelDBException, NoSuchAlgorithmException, InvalidProtocolBufferException {
+    Set<String> experimentRuns = new HashSet<>();
+    Map<Long, Map.Entry<RepositoryEntity, Map<String, Set<String>>>> blobs = new HashMap<>();
```

Comment:
There's no way I can parse 4 nested types in my head. Can we break this down?

Reply:
fixed